File: examples/mock_runtime_server.py
import os
from http import HTTPStatus

from fastapi import FastAPI, Request
from fastapi.responses import JSONResponse
import uvicorn

app = FastAPI(title="Keon MCP Demo Runtime Mock")


@app.get("/runtime/v1/status")
async def status():
    return {"status": "operational"}


@app.post("/runtime/v1/decide")
async def decide(request: Request):
    mode = os.getenv("DEMO_RUNTIME_MODE", "allow")
    body = await request.json()
    action = body.get("action", "")
    scope = body.get("resourceScope", "")

    if mode == "deny_mailbox_sent" and scope == "mailbox:sent":
        return {
            "success": True,
            "data": {
                "receiptId": "rcpt_dec_demo_deny",
                "decision": "deny",
                "policyHash": "sha256:demo-deny-policy",
                "policyId": "dlp.mail.summarize",
                "policyVersion": "2026-03-01",
                "reasonCode": "SENSITIVITY_LABEL_BLOCKED"
            }
        }

    if "deny" in action.lower():
        return {
            "success": True,
            "data": {
                "receiptId": "rcpt_dec_demo_deny",
                "decision": "deny",
                "policyHash": "sha256:demo-deny-policy",
                "policyId": "dlp.mail.summarize",
                "policyVersion": "2026-03-01",
                "reasonCode": "PURPOSE_MISMATCH"
            }
        }

    return {
        "success": True,
        "data": {
            "receiptId": "rcpt_dec_demo_allow",
            "decision": "allow",
            "policyHash": "sha256:demo-allow-policy",
            "policyId": "dlp.mail.summarize",
            "policyVersion": "2026-03-01"
        }
    }


@app.post("/runtime/v1/execute")
async def execute(request: Request):
    body = await request.json()
    return {
        "success": True,
        "data": {
            "executionReceiptId": "rcpt_exe_demo",
            "result": {
                "summary": "Governed summary generated.",
                "items_considered": body.get("parameters", {}).get("max_items", 0)
            }
        }
    }


@app.exception_handler(Exception)
async def unhandled(_: Request, ex: Exception):
    return JSONResponse(
        status_code=HTTPStatus.INTERNAL_SERVER_ERROR,
        content={"success": False, "error": {"message": str(ex)}},
    )


if __name__ == "__main__":
    port = int(os.getenv("PORT", "8080"))
    uvicorn.run(app, host="127.0.0.1", port=port, log_level="info")

File: tests/Keon.McpGateway.Tests/MockRuntimeServer.cs
using System.Net;
using System.Text.Json.Nodes;
using Microsoft.AspNetCore.Builder;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Hosting;
using Microsoft.Extensions.Hosting;

namespace Keon.McpGateway.Tests;

internal sealed class MockRuntimeServer : IAsyncDisposable
{
    private readonly WebApplication _app;
    private readonly Task _runTask;

    public MockRuntimeServer(RuntimeBehavior behavior)
    {
        var port = GetFreePort();
        BaseUrl = $"http://127.0.0.1:{port}";
        Behavior = behavior;

        var builder = WebApplication.CreateBuilder();
        builder.WebHost.UseUrls(BaseUrl);
        _app = builder.Build();

        _app.MapGet("/runtime/v1/status", async context =>
        {
            await context.Response.WriteAsJsonAsync(new { status = "operational" });
        });

        _app.MapPost("/runtime/v1/decide", async (HttpContext context) =>
        {
            if (Behavior.DecideStatusCode is not null)
            {
                context.Response.StatusCode = (int)Behavior.DecideStatusCode;
                await context.Response.WriteAsJsonAsync(new
                {
                    success = false,
                    error = new { message = "runtime decide unavailable" }
                });
                return;
            }

            var body = await context.Request.ReadFromJsonAsync<JsonObject>() ?? [];
            Behavior.DecideCorrelationIds.Add(body["correlationId"]?.GetValue<string>() ?? string.Empty);
            var action = body["action"]?.GetValue<string>() ?? string.Empty;
            var decision = action.Contains("deny", StringComparison.OrdinalIgnoreCase) ? "deny" : "allow";

            await context.Response.WriteAsJsonAsync(new
            {
                success = true,
                data = new
                {
                    receiptId = "rcpt_dec_mock",
                    decision,
                    policyHash = "sha256:1234567890abcdef",
                    policyId = "dlp.mail.summarize",
                    policyVersion = "2026-03-01",
                    reasonCode = decision == "deny" ? "SENSITIVITY_LABEL_BLOCKED" : (string?)null
                }
            });
        });

        _app.MapPost("/runtime/v1/execute", async (HttpContext context) =>
        {
            if (Behavior.ExecuteStatusCode is not null)
            {
                context.Response.StatusCode = (int)Behavior.ExecuteStatusCode;
                await context.Response.WriteAsJsonAsync(new
                {
                    success = false,
                    error = new { message = "runtime execute unavailable" }
                });
                return;
            }

            var body = await context.Request.ReadFromJsonAsync<JsonObject>() ?? [];
            Behavior.ExecuteCorrelationIds.Add(body["correlationId"]?.GetValue<string>() ?? string.Empty);
            await context.Response.WriteAsJsonAsync(new
            {
                success = true,
                data = new
                {
                    executionReceiptId = "rcpt_exe_mock",
                    result = new
                    {
                        summary = "done",
                        items_considered = 18
                    }
                }
            });
        });

        _runTask = _app.RunAsync();
    }

    public string BaseUrl { get; }
    public RuntimeBehavior Behavior { get; }

    public async ValueTask DisposeAsync()
    {
        await _app.StopAsync();
        await _runTask;
        await _app.DisposeAsync();
    }

    private static int GetFreePort()
    {
        using var listener = new System.Net.Sockets.TcpListener(IPAddress.Loopback, 0);
        listener.Start();
        var port = ((IPEndPoint)listener.LocalEndpoint).Port;
        listener.Stop();
        return port;
    }
}

internal sealed class RuntimeBehavior
{
    public HttpStatusCode? DecideStatusCode { get; set; }
    public HttpStatusCode? ExecuteStatusCode { get; set; }
    public List<string> DecideCorrelationIds { get; } = [];
    public List<string> ExecuteCorrelationIds { get; } = [];
}
